# **Import Library**

In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

# **Memuat Dataset dari Hasil Clustering**

In [3]:
transaksi_clustering= pd.read_csv("https://raw.githubusercontent.com/yovelakalista23/BMLP/refs/heads/main/hasil_clustering.csv")
transaksi_clustering.head()

TransactionAmount  CustomerAge  TransactionDuration  LoginAttempts  \
0              14.09         70.0                 81.0            1.0   
1             376.24         68.0                141.0            1.0   
2             126.29         19.0                 56.0            1.0   
3             184.50         26.0                 25.0            1.0   
4              13.45         26.0                198.0            1.0   

   AccountBalance  TransactionType_Credit  TransactionType_Debit  Channel_ATM  \
0         5112.21                       0                      1            1   
1        13758.91                       0                      1            1   
2         1122.35                       0                      1            0   
3         8569.06                       0                      1            0   
4         7429.40                       1                      0            0   

   Channel_Branch  Channel_Online  CustomerOccupation_Doctor  \
0               0               0                          1   
1               0               0                          1   
2               0               1                          0   
3               0               1                          0   
4               0               1                          0   

   CustomerOccupation_Engineer  CustomerOccupation_Retired  \
0                            0                           0   
1                            0                           0   
2                            0                           0   
3                            0                           0   
4                            0                           0   

   CustomerOccupation_Student  Cluster      PCA1      PCA2  
0                           0        2  0.814906  1.403180  
1                           0        0  2.712453 -0.654608  
2                           1        1 -1.782607 -0.609030  
3                           1        0  0.040627 -1.554162  
4                           1        1 -0.418018 -1.041885

# **Data Splitting**

In [10]:
from sklearn.model_selection import train_test_split

# Pisahkan fitur (X) dan target (y)
X = transaksi_clustering.drop(columns=['Cluster', 'PCA1', 'PCA2'])
y = transaksi_clustering['Cluster']  # Target: 'Cluster'

# Split data menjadi training dan testing set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Cek bentuk data setelah split
print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")


X_train shape: (1897, 14)
X_test shape: (475, 14)
y_train shape: (1897,)
y_test shape: (475,)


Memisahkan fitur X dan Y kemudian membagi dataset menjadi 2, yaitu training dataset dan testing dataset

# **Membangun model Klasifikasi**

## **a. Membangun Model Klasifikasi**

In [11]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

# Inisialisasi model
rf_model = RandomForestClassifier(random_state=42)
logreg_model = LogisticRegression(random_state=42, max_iter=1000)

Menggunakan algoritma Random Forest dan Logistic Regression

Kemudian Training Model

In [12]:
# Training model Random Forest
rf_model.fit(X_train, y_train)

# Training model Logistic Regression
logreg_model.fit(X_train, y_train)


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(max_iter=1000, random_state=42)

In [13]:
# Prediksi menggunakan Random Forest
y_pred_rf = rf_model.predict(X_test)

# Prediksi menggunakan Logistic Regression
y_pred_logreg = logreg_model.predict(X_test)

## **b. Evaluasi Model**

In [14]:
from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score, confusion_matrix, classification_report

# --- EVALUASI METRIK ---
# Random Forest
print("Evaluasi Random Forest:")
print(f"Accuracy: {accuracy_score(y_test, y_pred_rf):.4f}")
print(f"F1 Score: {f1_score(y_test, y_pred_rf, average='weighted'):.4f}")
print(f"Precision: {precision_score(y_test, y_pred_rf, average='weighted'):.4f}")
print(f"Recall: {recall_score(y_test, y_pred_rf, average='weighted'):.4f}")

print("\nClassification Report:\n")
print(classification_report(y_test, y_pred_rf))

# Logistic Regression
print("Evaluasi Logistic Regression:")
print(f"Accuracy: {accuracy_score(y_test, y_pred_logreg):.4f}")
print(f"F1 Score: {f1_score(y_test, y_pred_logreg, average='weighted'):.4f}")
print(f"Precision: {precision_score(y_test, y_pred_logreg, average='weighted'):.4f}")
print(f"Recall: {recall_score(y_test, y_pred_logreg, average='weighted'):.4f}")

print("\nClassification Report:\n")
print(classification_report(y_test, y_pred_logreg))

Evaluasi Random Forest:
Accuracy: 0.9663
F1 Score: 0.9664
Precision: 0.9678
Recall: 0.9663

Classification Report:

              precision    recall  f1-score   support

           0       0.93      0.99      0.96       162
           1       0.97      0.97      0.97       146
           2       1.00      0.94      0.97       167

    accuracy                           0.97       475
   macro avg       0.97      0.97      0.97       475
weighted avg       0.97      0.97      0.97       475

Evaluasi Logistic Regression:
Accuracy: 0.9642
F1 Score: 0.9642
Precision: 0.9644
Recall: 0.9642

Classification Report:

              precision    recall  f1-score   support

           0       0.96      0.98      0.97       162
           1       0.95      0.95      0.95       146
           2       0.98      0.96      0.97       167

    accuracy                           0.96       475
   macro avg       0.96      0.96      0.96       475
weighted avg       0.96      0.96      0.96       475



Evaluasi Random Forest:
- Accuracy: 0.9663
- F1 Score: 0.9664
- Precision: 0.9678
- Recall: 0.9663

Evaluasi Logistic Regression:
- Accuracy: 0.9642
- F1 Score: 0.9642
- Precision: 0.9644
- Recall: 0.9642

## **c. Tuning Model Klasifikasi**

Menggunakan GridSearchCV karena dataset yang digunakan tidak terlalu besar, sehingga prosesnya bisa optimal dan terkontrol

In [15]:
from sklearn.model_selection import GridSearchCV

In [16]:
# Parameter grid buat Random Forest
param_grid_rf = {
    'n_estimators': [100, 200, 300],  # jumlah pohon
    'max_depth': [None, 10, 20, 30],  # kedalaman pohon
    'min_samples_split': [2, 5, 10],  # minimum data untuk split
    'min_samples_leaf': [1, 2, 4]     # minimum data di tiap daun
}

# Model Random Forest
rf = RandomForestClassifier(random_state=42)

# Grid Search
grid_search_rf = GridSearchCV(estimator=rf, param_grid=param_grid_rf,
                              cv=5, n_jobs=-1, verbose=2, scoring='accuracy')

# Fit model
grid_search_rf.fit(X_train, y_train)

# Hasil terbaik
print("Best Parameters for Random Forest:", grid_search_rf.best_params_)
print("Best Score for Random Forest:", grid_search_rf.best_score_)


Fitting 5 folds for each of 108 candidates, totalling 540 fits
Best Parameters for Random Forest: {'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 200}
Best Score for Random Forest: 0.9783835578391891


Keterangan:

estimator=rf → model yang mau dituning.

param_grid=param_grid_rf → grid kombinasi hyperparameter.

cv=5 → pakai 5-fold cross-validation.

n_jobs=-1 → supaya pakai semua CPU core, biar cepet.

verbose=2 → biar bisa liat progress di terminal.

scoring='accuracy' → kita optimalkan accuracy.

In [17]:
# Parameter grid buat Logistic Regression
param_grid_logreg = {
    'C': [0.01, 0.1, 1, 10, 100],        # regularization strength
    'penalty': ['l1', 'l2'],              # jenis regularization
    'solver': ['liblinear']               # solver yang support l1 dan l2
}

# Model Logistic Regression
logreg = LogisticRegression(random_state=42, max_iter=1000)

# Grid Search
grid_search_logreg = GridSearchCV(estimator=logreg, param_grid=param_grid_logreg,
                                  cv=5, n_jobs=-1, verbose=2, scoring='accuracy')

# Fit model
grid_search_logreg.fit(X_train, y_train)

# Hasil terbaik
print("Best Parameters for Logistic Regression:", grid_search_logreg.best_params_)
print("Best Score for Logistic Regression:", grid_search_logreg.best_score_)


Fitting 5 folds for each of 10 candidates, totalling 50 fits
Best Parameters for Logistic Regression: {'C': 10, 'penalty': 'l1', 'solver': 'liblinear'}
Best Score for Logistic Regression: 0.9899819469518123


Penjelasan:

param_grid=param_grid_logreg: Grid hyperparameter yang sudah disiapkan.

cv=5: 5-fold cross-validation, membagi data menjadi 5 bagian untuk validasi silang.

n_jobs=-1: Agar semua core CPU digunakan untuk mempercepat proses pencarian.

verbose=2: Agar kita bisa melihat progres grid search.

scoring='accuracy': Kita mengoptimalkan berdasarkan accuracy.

## **d. Evaluasi Model Setelah Melakukan Tuning**

In [18]:
best_model_rf = grid_search_rf.best_estimator_  # Random Forest
best_model_logreg = grid_search_logreg.best_estimator_  # Logistic Regression

In [19]:
y_pred_rf_tuned = best_model_rf.predict(X_test)  # Prediksi dengan Random Forest
y_pred_logreg_tuned = best_model_logreg.predict(X_test)  # Prediksi dengan Logistic Regression

In [20]:
accuracy_rf_tuned = accuracy_score(y_test, y_pred_rf_tuned)  # Random Forest
f1_rf_tuned = f1_score(y_test, y_pred_rf_tuned, average='weighted')  # Random Forest
precision_rf_tuned = precision_score(y_test, y_pred_rf_tuned, average='weighted')
recall_rf_tuned = recall_score(y_test, y_pred_rf_tuned, average='weighted')

f1_logreg_tuned = f1_score(y_test, y_pred_logreg_tuned, average='weighted')  # Logistic Regression
accuracy_logreg_tuned = accuracy_score(y_test, y_pred_logreg_tuned)  # Logistic Regression
precision_logreg_tuned = precision_score(y_test, y_pred_logreg_tuned, average='weighted')
recall_logreg_tuned = recall_score(y_test, y_pred_logreg_tuned, average='weighted')


In [29]:
print(f"Random Forest Accuracy (tuned): {accuracy_rf_tuned:.4f}")
print(f"Random Forest F1-Score (tuned): {f1_rf_tuned:.4f}")
print(f"Random Forest Precision (tuned): {precision_rf_tuned:.4f}")
print(f"Random Forest Recall (tuned): {recall_rf_tuned:.4f}")

Random Forest Accuracy (tuned): 0.9726
Random Forest F1-Score (tuned): 0.9727
Random Forest Precision (tuned): 0.9734
Random Forest Recall (tuned): 0.9726


In [30]:
print(f"Logistic Regression Accuracy (tuned): {accuracy_logreg_tuned:.4f}")
print(f"Logistic Regression F1-Score (tuned): {f1_logreg_tuned:.4f}")
print(f"Logistic Regression Precision (tuned): {precision_logreg_tuned:.4f}")
print(f"Logistic Regression Recall (tuned): {recall_logreg_tuned:.4f}")

Logistic Regression Accuracy (tuned): 0.9747
Logistic Regression F1-Score (tuned): 0.9748
Logistic Regression Precision (tuned): 0.9757
Logistic Regression Recall (tuned): 0.9747


## **e. Analisis Hasil Evaluasi Setelah Tuning**

Peningkatan Kinerja:

Setelah tuning, kedua model mengalami peningkatan performa yang signifikan.

**Random Forest mengalami peningkatan di semua metrik:**

- Accuracy meningkat dari 0.9663 menjadi 0.9726.

- F1-Score meningkat dari 0.9664 menjadi 0.9727.

- Precision meningkat dari 0.9678 menjadi 0.9734.

- Recall meningkat dari 0.9663 menjadi 0.9726.

**Logistic Regression juga menunjukkan peningkatan:**

- Accuracy meningkat dari 0.9642 menjadi 0.9747.

- F1-Score meningkat dari 0.9642 menjadi 0.9748.

- Precision meningkat dari 0.9644 menjadi 0.9757.

- Recall meningkat dari 0.9642 menjadi 0.9747.


Impak Tuning Hyperparameter:

Tuning hyperparameter jelas memberikan peningkatan yang signifikan dalam performa kedua model. Ini membuktikan pentingnya optimasi parameter untuk mencapai hasil terbaik.

Logistic Regression vs Random Forest:

Logistic Regression memiliki Accuracy, F1-Score, Precision, dan Recall yang sedikit lebih tinggi dibandingkan Random Forest setelah tuning.

Ini menunjukkan bahwa Logistic Regression lebih optimal untuk dataset ini setelah tuning, meskipun keduanya menunjukkan performa yang sangat baik.